### PR TITLE
Document how to treat inhomogeneous Dirichlet values with MatrixFree

### LINKED ISSUE #5339: Document how to handle non-homogeneous Dirichlet constraints in step-37
We should add the code linked in the google group [1] in the possibilities for extensions section of step-37

[1] https://groups.google.com/forum/#!topic/dealii/g2bU8SCK24g

### PR DESCRIPTION
**Pull request by kronbichler**
Fixes #5339.

Since there are (at least) two possible ways of doing this, I described them both here.

### CODE REVIEW

**Comment by luca-heltai** [comment]
The other possibility, is to use constrained_linear_operator to wrap only the Dirichlet constraints...

https://www.dealii.org/developer/doxygen/deal.II/group__constraints.html#ga59c2f70370213d436cb9fade9d813478

File: examples/step-37/doc/results.dox [hunk 1]
```diff
@@ -409,3 +411,229 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+assumed that the degrees of freedom are numbered such that the ones subject to
+a Dirichlet condition are all appended to the end of the list of degrees of
+freedom. We then insert this solution representation into the weak form,
+e.g. the Laplacian shown above, and move the known quantities to the right
+hand side:
+@f{eqnarray*}
+(\varphi_i, u_h)_\Omega &=& (\varphi_i, f)_\Omega \Rightarrow \\
+\sum_{j=1}^{n_\mathrm{dofs}-n_\mathrm{dofs,D}}(\varphi_i,\varphi_j)_\Omega u_j =
+(\varphi_i, f)_\Omega
+-\sum_{j=n_\mathrm{dofs}-n_\mathrm{dofs,D}+1}^{n_\mathrm{dofs}} (\varphi_i,\varphi_j)_\Omega g_j.
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: maybe add space after `_Omega` so that `g_j` is not that close and it's easier to read. sabe above for `u_j`.

File: examples/step-37/doc/results.dox [hunk 2]
```diff
@@ -409,3 +411,229 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+
+In the implementation in deal.II, the integrals $(\varphi_i,\varphi_j)_\Omega$
+on the right hand side are already contained in the local matrix contributions
+we assemble on each cell. When using
+ConstraintMatrix::distributed_local_to_global as first described in the step-6
+and step-7 tutorial programs, we can account for the contribution of
+inhomogeneous constraints $j$ by multiplying the columns $j$ and rows $i$ of
+the local matrix according to the integrals (\varphi_i, \varphi_j)_\Omega$ by
+the inhomogeneities and subtracting the resulting from the position $i$ in the
+global vector, see also the @ref constraints module. In essence, we use some
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: in the global **right-hand-side** vector

File: examples/step-37/doc/results.dox [hunk 3]
```diff
@@ -409,3 +412,232 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+  solution = 0;
+  constraints.distribute(solution);
+@endcode
+
+We could then pass the vector @p solution to the @p LaplaceOperator::vmult_add()
+function and add the new contribution to the @p system_rhs vector that gets
+filled in the @p LaplaceProblem::assemble_rhs() function. However, this idea does
+not work because the vmult() functions eliminates the entries filled into
+Dirichlet-constrained degrees of freedom on the fly by construction. Instead,
+we could select one of two following strategies.
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: i think the already apply internally in `project()` a third way: get RHS contributions from constraints by doing vmult, add to RHS and then through that into solver:
https://github.com/dealii/dealii/blob/master/include/deal.II/numerics/vector_tools.templates.h#L1304-L1326
This does not appear to be any of the two options below.

File: examples/step-37/doc/results.dox (hunk 1 above)
Comment (kronbichler) [reply examples/step-37/doc/results.dox]: That's a good idea indeed, yes. I will adjust.

File: examples/step-37/doc/results.dox [hunk 4]
```diff
@@ -409,3 +412,232 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+
+<h5> Use FEEvaluation::read_dof_values_plain() to avoid resolving constraints </h5>
+
+The class FEEvaluation has a facility that is addresses precisely this
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: ~~is~~

File: examples/step-37/doc/results.dox [hunk 5]
```diff
@@ -409,3 +412,232 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+  system_rhs.compress(VectorOperation::add);
+}
+@endcode
+
+In this code, we replaced the FEEvaluation::read_dof_values() function for the
+tentative solution vector by FEEvaluation::read_dof_values_plain() that
+ignores all constraints. Due to this setup, we must make sure that other
+constraints, e.g. by hanging nodes, are correctly distributed to the input
+vector already as they are not resolved as in
+FEEvaluation::read_dof_values(). Inside the loop, we then evaluate the
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: read_dof_values_plain?

File: examples/step-37/doc/results.dox (hunk 3 above)
Comment (davydden) [reply examples/step-37/doc/results.dox]: i went through the two options now and I think what we do in `project_matrix_free` (originally written by @masterleinad ) is a more concise correct third option.

File: examples/step-37/doc/results.dox [hunk 6]
```diff
@@ -409,3 +411,229 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+The presented program assumes homogeneous Dirichlet boundary conditions. When
+going to non-homogeneous conditions, the situation is a bit more intricate. To
+understand how to implement such a setting, let us first recall how these
+arise in the mathematical formulation and how they are implemented in a
+matrix-based variant. In essence, an inhomogeneous Dirichlet condition sets
+some of the nodal values in the solution to given values rathern than
+determining them through the variational principles,
+@f{eqnarray*}
+u_h(\mathbf{x}) = \sum_{i=1}^{n_\mathrm{dofs}} \varphi_i(\mathbf{x}) u_i =
+\sum_{i=1}^{n_\mathrm{dofs}-n_\mathrm{dofs,D}} \varphi_i(\mathbf{x}) u_i +
```
[the remark this answers is not in the document]
Comment (davydden) [reply examples/step-37/doc/results.dox]: ps. updated the comment to use `\setminus`

File: examples/step-37/doc/results.dox (hunk 3 above)
Comment (kronbichler) [reply examples/step-37/doc/results.dox]: I also checked this (and I remember having reviewed it), but it does not work. I just tried to change `ZeroFunction` into `ConstantFunction<dim>(1.)` in step-37 and it does ignore that. The reason is explained right in this block: The standard implementation of the `vmult()` method does set the constrained degrees of freedom to zero while reading them out of the vector. Thus, in absence of hanging node constraints it simply ignores those. I'm not sure how well we tested the `project()` function but I don't understand how it can work. The "correct" way of doing the `project` is to use the second option I describe.

Comment (davydden) [reply examples/step-37/doc/results.dox]: ok, but I still have a third options as a suggestion: use two different instances of `LaplaceOperator` and `MatrixFree` objects, one with constraints and one without. To me the advantage is that you can use exactly the same code in `vmult` etc, so no need to get special implementation only to make BC working.

Comment (kronbichler) [reply examples/step-37/doc/results.dox]: But isn't that just a variant of option two? (Except that I suggest to add a second `DoFHandler` object and create a second `LaplaceOperator` instance - it could of course also be an independent `MatrixFree` object. I can add it if you want there, or even show that as the "option 2" instead if you think it is easier to understand than adding a second DoFHandler/ConstraintMatrix pair.

File: examples/step-37/doc/results.dox (hunk 6 above)
[the remark this answers is not in the document]
Comment (kronbichler) [reply examples/step-37/doc/results.dox]: I did that already, thanks for the excellent suggestion. :+1:

File: examples/step-37/doc/results.dox (hunk 3 above)
Comment (davydden) [reply examples/step-37/doc/results.dox]: I though it could be more general if the system to be solved is block system. Then presumably you already reserve blocks for different blocks in the operator and it would be more difficult to introduce extra blocks only for BC.

I suggest to move the third option to `MatrixFreeOperators::Base` with a super short discussion and link to step-37.

File: examples/step-37/doc/results.dox [hunk 7]
```diff
@@ -409,3 +412,235 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+matrix-based variant. In essence, an inhomogeneous Dirichlet condition sets
+some of the nodal values in the solution to given values rather than
+determining them through the variational principles,
+@f{eqnarray*}
+u_h(\mathbf{x}) = \sum_{i\in \mathcal N} \varphi_i(\mathbf{x}) u_i =
+\sum_{i\in \mathcal N \setminus \mathcal N_D} \varphi_i(\mathbf{x}) u_i +
+\sum_{i\in \mathcal N_D} \varphi_i(\mathbf{x}) g_i,
+@f}
+where $u_i$ denotes the nodal values of the solution and $\mathcal N$ denotes
+the set of all nodes. The set $\mathcal N_D\subset \mathcal N$ is the subset
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: should be `$\mathcal N_D$`

Comment (davydden) [reply examples/step-37/doc/results.dox]: sorry, what's wrong with me today 😄

Comment (davydden) [reply examples/step-37/doc/results.dox]: again I read it different to what is written: $\mathcal N_D \setminus \mathcal N$

File: examples/step-37/doc/results.dox [hunk 8]
```diff
@@ -409,3 +412,235 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+N_D$. In this formula, where we have assumed that the degrees of freedom are
+numbered such that the ones subject to a Dirichlet condition are all appended
+to the end of the list of degrees of freedom. We then insert this solution
+representation into the weak form, e.g. the Laplacian shown above, and move
+the known quantities to the right hand side:
+@f{eqnarray*}
+(\varphi_i, u_h)_\Omega &=& (\varphi_i, f)_\Omega \quad \Rightarrow \\
+\sum_{j\in \mathcal N \setminus \mathcal N_D}(\varphi_i,\varphi_j)_\Omega \, u_j &=&
+(\varphi_i, f)_\Omega
+-\sum_{j\in \mathcal N_D} (\varphi_i,\varphi_j)_\Omega g_j.
```
Comment (davydden) [inline examples/step-37/doc/results.dox]: i think you wanted to add space: `(\varphi_i,\varphi_j)_\Omega\,\,g_j`

Comment (kronbichler) [reply examples/step-37/doc/results.dox]: You didn't see that, did you? :smiley: (I forgot that second one, I've just pushed the correction.)

File: examples/step-37/doc/results.dox (hunk 3 above)
Comment (kronbichler) [reply examples/step-37/doc/results.dox]: So how about I add a sentence in the second option that this could also be done via a separate `MatrixFree`/`LaplaceOperator` pair and then extend `MatrixFreeOperators::Base`?

Comment (davydden) [reply examples/step-37/doc/results.dox]: that sounds good to me.

Comment (kronbichler) [reply examples/step-37/doc/results.dox]: OK, I've added the discussion to `MatrixFreeOperators::Base`. Note that I changed the order of the two existing paragraphs to have the third option extend the second one (which is the selection of another block).

File: include/deal.II/matrix_free/operators.h [hunk 9]
```diff
@@ -141,22 +158,26 @@ namespace MatrixFreeOperators
...
-   * be sure that MatrixFree uses the same integer indexing to the different
-   * blocks, one needs to put all three blocks into the same MatrixFree
-   * object. However, when solving a linear system the operators involved
-   * either address the first two in the Stokes solver, or the last one for
-   * the temperature solver. In the former case, a BlockVector of two
-   * components would be selected with a vector selecting the blocks {0, 1} in
-   * MatrixFree, whereas in the latter, a non-block vector selecting the block
-   * {2} would be used.
+   * For systems of PDEs where the different blocks of MatrixFree are
+   * associated different physical components of the equations, adding another
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: associated with?

File: include/deal.II/matrix_free/operators.h [hunk 10]
```diff
@@ -141,22 +158,26 @@ namespace MatrixFreeOperators
...
+   * block with a different ConstraintMatrix argument solely for the purpose
+   * of boundary conditions might lead to cumbersome index handling. Instead,
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: yes, that's exactly what I though^^^

File: include/deal.II/matrix_free/operators.h [hunk 11]
```diff
@@ -141,22 +158,26 @@ namespace MatrixFreeOperators
...
+   * in the results section of the step-37 tutorial program):
+   *
+   * @code
+   * matrix_free_inhomogeneous.reinit(dof_handler, constraints_no_dirichlet,
+   *                                  quadrature, additional_data);
+   * operator_inhomogeneous.initialize(matrix_free_inhomogeneous, selected_block);
+   * LinearAlgebra::distributed::Vector<double> inhomogeneity;
+   * matrix_free_inhomogeneous.initialize_dof_vector(inhomogeneity);
+   * constraints_with_dirichlet.distribute(inhomogeneity);
+   * laplace_operator_inhomogeneous.vmult(system_rhs, inhomogeneity);
```
Comment (davydden) [inline include/deal.II/matrix_free/operators.h]: `operator_inhomogeneous` ? (that's what you used above)

**Comment by davydden** [comment]
@kronbichler @masterleinad :

A few more thoughts on the constraints discussion. Looks like this comment above

> The standard implementation of the vmult() method does set the constrained degrees of freedom to zero while reading them out of the vector. Thus, in absence of hanging node constraints it simply ignores those.

implies that we set constrained dofs to zero upon reading, whereas the added text:

> For non-homogeneous Dirichlet values, we do want to skip the
 resolution of the (Dirichlet) constraints upon reading the data from the
 vector @p solution.

to me says that they are resolved (so Dirichlet are set to actual numbers). I believe the latter is correct. 

If so, then `solution=0` and reading from it via `phi.read_dof_values()` we should get values with Dirichlet constraints set? 

The comes standard steps of `vmult()` and finally `phi.distribute_local_to_global()`. But this should be fine as constrained rows are irrelevant, whereas non-constrained rows with non-zero entries after `vmult` shall be distributed correctly without any modifications.

Bottom line: I am still puzzled why our current `project_matrix_free` is wrong https://github.com/dealii/dealii/blob/master/include/deal.II/numerics/vector_tools.templates.h#L1304-L1326 ?

**Comment by kronbichler** [comment]
> > The standard implementation of the vmult() method does set the constrained degrees of freedom to zero while reading them out of the vector. Thus, in absence of hanging node constraints it simply ignores those.
>
> implies that we set constrained dofs to zero upon reading,
> > For non-homogeneous Dirichlet values, we do want to skip the resolution of the (Dirichlet) constraints upon reading the data from the vector @p solution.
>
> to me says that they are resolved (so Dirichlet are set to actual numbers). I believe the latter is correct.

They are not in contradiction, just non expressed in a precise manner (I will update in a minute). The `read_dof_values` does resolve the **homogeneous** constraints, so any inhomogeneous Dirichlet value will be set to zero in the `read_dof_values` call. Thus, the Dirichlet constraints do not get set. And we cannot set them, because we must assume that the user wants us to define a linear operator, not an affine one.

> Also IIRC we have some tests with non-homogeneous constraints for project, which were not broken by switching to matrix-free parallel code path.

I let @masterleinad answer this more precisely, but all I remember is that we hardly ever set constraints in the `VectorTools::project` but rather let the method adjust all values freely, including the ones that later get Dirichlet values. Now regardless the actual state of the tests, no doubt we definitely should have non-homogeneous constraints in the `project` tests to check this code path.

File: examples/step-37/doc/results.dox [hunk 12]
```diff
@@ -409,3 +412,240 @@ number of threads in the MPI_InitFinalize data structure in the main function,
...
+numbered such that the ones subject to a Dirichlet condition are all appended
+to the end of the list of degrees of freedom. We then insert this solution
+representation into the weak form, e.g. the Laplacian shown above, and move
+the known quantities to the right hand side:
+@f{eqnarray*}
+(\varphi_i, u_h)_\Omega &=& (\varphi_i, f)_\Omega \quad \Rightarrow \\
+\sum_{j\in \mathcal N \setminus \mathcal N_D}(\varphi_i,\varphi_j)_\Omega \, u_j &=&
+(\varphi_i, f)_\Omega
+-\sum_{j\in \mathcal N_D} (\varphi_i,\varphi_j)_\Omega\, g_j.
+@f}
```
Comment (masterleinad) [inline examples/step-37/doc/results.dox]: Don't you want to consider here a Laplace matrix instead of a mass matrix, i.e. 
```
(\nabla \varphi_i, \nabla \u_h) &=& (\varphi_i, f)_\Omega
```
?

Comment (kronbichler) [reply examples/step-37/doc/results.dox]: Of course, thanks!

**Comment by davydden** [comment]
> The does resolve the homogeneous constraints, so any inhomogeneous Dirichlet value will be set to zero in the read_dof_values call. Thus, the Dirichlet constraints do not get set.

I see now, thanks for explaining @kronbichler . Makes perfect sense. Actually documentation of `read_dof_values` already says that, I just never checked it, i think.
